File: app/src/main/java/remix/myplayer/ui/activity/SongChooseActivity.java
package remix.myplayer.ui.activity;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.Loader;
import android.os.Bundle;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;
import java.util.List;
import org.jetbrains.annotations.Nullable;
import remix.myplayer.R;
import remix.myplayer.bean.mp3.Song;
import remix.myplayer.databinding.ActivitySongChooseBinding;
import remix.myplayer.db.room.DatabaseRepository;
import remix.myplayer.misc.asynctask.AppWrappedAsyncTaskLoader;
import remix.myplayer.misc.interfaces.LoaderIds;
import remix.myplayer.request.network.RxUtil;
import remix.myplayer.theme.ThemeStore;
import remix.myplayer.ui.adapter.SongChooseAdapter;
import remix.myplayer.util.MediaStoreUtil;
import remix.myplayer.util.ToastUtil;

/**
 * @ClassName SongChooseActivity
 * @Description 新建列表后添加歌曲
 * @Author Xiaoborui
 * @Date 2016/10/21 09:34
 */

public class SongChooseActivity extends LibraryActivity<Song, SongChooseAdapter> {
  private ActivitySongChooseBinding binding;

  public static final String TAG = SongChooseActivity.class.getSimpleName();
  public static final String EXTRA_NAME = "PlayListName";
  public static final String EXTRA_ID = "PlayListID";

  private int mPlayListID;
  private String mPlayListName;

  public static void start(Activity activity, int playListId, String playListName) {
    Intent intent = new Intent(activity, SongChooseActivity.class);
    intent.putExtra(EXTRA_ID, playListId);
    intent.putExtra(EXTRA_NAME, playListName);
    activity.startActivity(intent);
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    binding = ActivitySongChooseBinding.inflate(getLayoutInflater());
    setContentView(binding.getRoot());

    mPlayListID = getIntent().getIntExtra(EXTRA_ID, -1);
    if (mPlayListID <= 0) {
      ToastUtil.show(this, R.string.add_error, Toast.LENGTH_SHORT);
      return;
    }
    mPlayListName = getIntent().getStringExtra(EXTRA_NAME);

    mAdapter = new SongChooseAdapter(R.layout.item_song_choose, isValid -> {
      binding.confirm.setAlpha(isValid ? 1.0f : 0.6f);
      binding.confirm.setClickable(isValid);
    });

    binding.recyclerview.setLayoutManager(new LinearLayoutManager(this));
    binding.recyclerview.setAdapter(mAdapter);
    binding.recyclerview.setItemAnimator(new DefaultItemAnimator());
    binding.confirm.setAlpha(0.6f);

    binding.header.setBackgroundColor(ThemeStore.getMaterialPrimaryColor());
    for (TextView view : new TextView[]{binding.confirm, binding.cancel, binding.title}) {
      view.setTextColor(ThemeStore.getTextColorPrimaryReverse());
    }

    binding.confirm.setOnClickListener(this::onClick);
    binding.cancel.setOnClickListener(this::onClick);
  }

  public void onClick(View v) {
    int id = v.getId();
    if (id == R.id.cancel) {
      finish();
    } else if (id == R.id.confirm) {
      if (mAdapter.getCheckedSong() == null || mAdapter.getCheckedSong().size() == 0) {
        ToastUtil.show(this, R.string.choose_no_song);
        return;
      }
      DatabaseRepository.getInstance()
          .insertToPlayList(mAdapter.getCheckedSong(), mPlayListID)
          .compose(RxUtil.applySingleScheduler())
          .subscribe(num -> {
            ToastUtil.show(mContext, getString(R.string.add_song_playlist_success, num,
                mPlayListName));
            finish();
          }, throwable -> finish());
    }
  }

  @Override
  protected int getLoaderId() {
    return LoaderIds.ACTIVITY_SONGCHOOSE;
  }


  @Override
  protected Loader<List<Song>> getLoader() {
    return new AsyncSongLoader(mContext);
  }

  @Override
  public void saveSortOrder(@Nullable String sortOrder) {

  }

  private static class AsyncSongLoader extends AppWrappedAsyncTaskLoader<List<Song>> {

    private AsyncSongLoader(Context context) {
      super(context);
    }

    @Override
    public List<Song> loadInBackground() {
      return MediaStoreUtil.getAllSong();
    }
  }

}

File: app/src/main/java/remix/myplayer/ui/activity/SongChooseActivity.kt
package remix.myplayer.ui.activity

import android.content.Context
import android.content.Intent
import android.content.Loader
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import remix.myplayer.R
import remix.myplayer.bean.mp3.Song
import remix.myplayer.databinding.ActivitySongChooseBinding
import remix.myplayer.db.room.DatabaseRepository.Companion.getInstance
import remix.myplayer.misc.asynctask.AppWrappedAsyncTaskLoader
import remix.myplayer.misc.interfaces.LoaderIds
import remix.myplayer.misc.interfaces.OnSongChooseListener
import remix.myplayer.util.RxUtil
import remix.myplayer.theme.ThemeStore.materialPrimaryColor
import remix.myplayer.theme.ThemeStore.textColorPrimaryReverse
import remix.myplayer.ui.adapter.SongChooseAdapter
import remix.myplayer.util.MediaStoreUtil.getAllSong
import remix.myplayer.util.ToastUtil

/**
 * @ClassName SongChooseActivity
 * @Description 新建列表后添加歌曲
 * @Author Xiaoborui
 * @Date 2016/10/21 09:34
 */
class SongChooseActivity : LibraryActivity<Song, SongChooseAdapter>() {
  lateinit var binding: ActivitySongChooseBinding

  private var mPlayListID = 0
  private var mPlayListName: String? = null
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    binding = ActivitySongChooseBinding.inflate(layoutInflater)
    setContentView(binding.root)
    mPlayListID = intent.getIntExtra(EXTRA_ID, -1)
    if (mPlayListID <= 0) {
      ToastUtil.show(this, R.string.add_error, Toast.LENGTH_SHORT)
      return
    }
    mPlayListName = intent.getStringExtra(EXTRA_NAME)
    adapter = SongChooseAdapter(R.layout.item_song_choose, OnSongChooseListener { isValid: Boolean ->
      binding.confirm.alpha = if (isValid) 1.0f else 0.6f
      binding.confirm.isClickable = isValid
    })
    binding.recyclerview.layoutManager = LinearLayoutManager(this)
    binding.recyclerview.adapter = adapter
    binding.recyclerview.itemAnimator = DefaultItemAnimator()
    binding.confirm.alpha = 0.6f
    binding.header.setBackgroundColor(materialPrimaryColor)
    for (view in arrayOf(binding.confirm, binding.cancel, binding.title)) {
      view.setTextColor(textColorPrimaryReverse)
    }
    binding.confirm.setOnClickListener { v: View -> onClick(v) }
    binding.cancel.setOnClickListener { v: View -> onClick(v) }
  }

  fun onClick(v: View) {
    val id = v.id
    if (id == R.id.cancel) {
      finish()
    } else if (id == R.id.confirm) {
      if (adapter?.checkedSong?.size == 0) {
        ToastUtil.show(this, R.string.choose_no_song)
        return
      }
      getInstance()
          .insertToPlayList(adapter!!.checkedSong, mPlayListID.toLong())
          .compose(RxUtil.applySingleScheduler())
          .subscribe({ num: Int? ->
            ToastUtil.show(this, getString(R.string.add_song_playlist_success, num,
                mPlayListName))
            finish()
          }) { throwable: Throwable? -> finish() }
    }
  }

  override val loaderId: Int = LoaderIds.ACTIVITY_SONGCHOOSE

  override val loader: Loader<List<Song>>
    get() = AsyncSongLoader(this)

  override fun saveSortOrder(sortOrder: String) {

  }

  private class AsyncSongLoader(context: Context) : AppWrappedAsyncTaskLoader<List<Song>>(context) {
    override fun loadInBackground(): List<Song> {
      return getAllSong()
    }
  }

  companion object {
    val TAG = SongChooseActivity::class.java.simpleName
    const val EXTRA_NAME = "PlayListName"
    const val EXTRA_ID = "PlayListID"
    fun start(context: Context, playListId: Int, playListName: String?) {
      val intent = Intent(context, SongChooseActivity::class.java)
      intent.putExtra(EXTRA_ID, playListId)
      intent.putExtra(EXTRA_NAME, playListName)
      context.startActivity(intent)
    }
  }
}